R/calculateOverlap.R

```diff
@@ -82,55 +82,72 @@ setMethod("calculateOverlap", signature = c(x = "SummarizedExperiment"),
         }
         detection <- as.numeric(detection)
         ########################### INPUT CHECK END ############################
-        # Get assay
-        assay <- assay(x, assay.type)
-        
+        mat <- assay(x, assay.type)
+        res <- getOverlap(mat, detection, ...)
+        return(res)
+    }
+)
+
+#' @rdname getOverlap
+#' @export
+setMethod("getOverlap", signature = c(x = "ANY"),
+    function(x,detection = 0, ...){
+        ############################# INPUT CHECK ##############################
+        # Check detection
+        if (!.is_numeric_string(detection)) {
+          stop("'detection' must be a single numeric value or coercible to ",
+               "one.",
+               call. = FALSE)
+        }
+        detection <- as.numeric(detection)
```

Comment:
Same here. Can we convert this to internal function and call it from getDissimilarity?

Reply:
The problem is similar to this in addAlpha https://github.com/microbiome/mia/pull/612/files

When we are just adding new components, it makes the code harder to maintain.

We could just have getDissimilarity that calls

- .get_unifrac
- .get_overlap
- .get_jsd

Those internal functions should have as little as possible functionality (they should have only method-specific functionality.


In fact, I think getDissimilarity already have most of the functionality needed. It is only missing tree functionality. (Is the tree that we use from rowTree()/colTree() slot or is it directly provided by user). You could implement it here .calculate_dissimilarity (as it should be hidden because it is not used by every method).

Check addAlpha for idea. Let's discuss more if you did not get the idea